src/main.rs

```diff
@@ -57,7 +58,19 @@ fn run_target_program(elf_path: &Path, chip_name: &str, opts: &cli::Opts) -> any
     let mut sess = if opts.connect_under_reset {
         probe.attach_under_reset(probe_target)?
     } else {
-        probe.attach(probe_target)?
+        let probe_attach = probe.attach(probe_target);
+        if let Err(probe_rs::Error::Probe(ProbeSpecific(e))) = &probe_attach {
+            if e.to_string().contains("JtagNoDeviceConnected") {
```

Comment:
It would be better to downcast to the concrete error type and inspect that. If probe-rs doesn't expose that type, please add a `// FIXME` comment explaining that.

Reply:
`e` is `Box<dyn std::error::Error + Send + Sync>`, can I downcast it?

`dbg!` tells me

```
e = CommandFailed(
    JtagNoDeviceConnected,
)
```

And using `cargo doc` I cannot find `CommandFailed` nor `JtagNoDeviceConnected`.

Reply:
Searching through the repo I found the `enum StlinkError::CommandFailed(...)` ([link](https://github.com/probe-rs/probe-rs/blob/b7e7cfd667a0a85fe012017f3cf865038001970e/probe-rs/src/probe/stlink/mod.rs#L1122)). This is only `pub(crate)` and therefore not accessible for us. If I understand it correctly we therefore can't downcast it.

@jonas-schievink What do you think is the best way to move forward? Ask them to expose the type?

Reply:
> Ask them to expose the type?

Yes, that is needed to implement this more cleanly.

For now, we can use the `.to_string()` solution, since this is just a diagnostics improvement anyways. There should be at least a `// FIXME` comment in the code explaining this though.

Reply:
Comment suggestion:

```rust
// FIXME Using `to_string().contains(...)` is a workaround as the concrete type of
// `e` is not public and therefore does not allow downcasting.
```